### PR TITLE
Ncgames 10

### PR DESCRIPTION
**Pull request by Wiggy93**
Completed happy and sad path testing

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -213,131 +213,289 @@ describe('PATCH commands', () => {
...
-         expect(body).toEqual({"message" : "Invalid path provided - please try again"});
+describe('Get /api/reviews Queries', () => {
+    test('200: should return the reviews with the category dexterity', () => {
+        return request(app).get("/api/reviews?category=dexterity")
+        .expect(200)
+        .then(({body})=>{
+            const reviews = body.reviews;
+          
+            expect(reviews).toHaveLength(1);
+            expect(reviews[0]).toEqual({
```
Comment (Rosem10) [inline __tests__/app.test.js]: Please check that all of the reviews have the expected category using something like a ForEach (you don't have to check all of the other keys though).

Comment (Rosem10) [reply __tests__/app.test.js]: I just realised that you're testing for each individual category. You should only have one test for "only returns reviews for the category passed", then use one of the categories to prove this. What if more categories are added later? Would you have to write a test for each one? That would be really labour intensive and unnecessary as we would already know that this functionality works. Also doing it this way is not very DRY.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -213,131 +213,289 @@ describe('PATCH commands', () => {
...
+            const reviews = body.reviews
+            expect(reviews).toHaveLength(0);
+        })
     });
 
-    test('400: GET followed by a comment on an invalid review datatype should return a message', () => {
-        return request(app)
-        .get("/api/reviews/bananas/comments")
-        .expect(400)
+    test('200: should return all reviews sorted by review_id ascending', () => {
```
Comment (Rosem10) [inline __tests__/app.test.js]: Same for this, you only need one test to prove all sortBys

File: __tests__/app.test.js [hunk 3]
```diff
@@ -213,131 +213,289 @@ describe('PATCH commands', () => {
...
+                expect(review.category).toBe("social deduction")
+            })
         })
     });
 
-    test('400: GET follow by invalid review id ie resource that doesn\'t exist should return a message', () => {
-        return request(app)
-        .get("/api/reviews/99999")
-        .expect(404)
+    test('200: should return no reviews with category children\'s games', () => {
```
Comment (Rosem10) [inline __tests__/app.test.js]: This one can stay, but please name it more generically - we're not testing specifically for Children's Games (even though we are using it as an example in our tests) - we're testing more generally for an empty array when a category has no reviews.

File: __tests__/app.test.js [hunk 4]
```diff
@@ -213,131 +213,289 @@ describe('PATCH commands', () => {
...
 
-    test('404 status: POST using a username not in the database should return an error message', () => {
-        return request(app)
-        .post("/api/reviews/1/comments")
-        .send({
-            username: "Alex",
-            body: "This body shouldn't be here!",
-        })
-        .expect(404)
+    test('200: should return all reviews sorted by votes ascending', () => {
```
Comment (Rosem10) [inline __tests__/app.test.js]: Please also check for descending, and also test for the expected default behaviour is no order is specified.

File: __tests__/app.test.js (hunk 2 above)
Comment (Rosem10) [reply __tests__/app.test.js]: Please also write a test for default behaviour if no sort_by is given.

File: models/app.model.js [hunk 5]
```diff
@@ -7,16 +7,34 @@ const fetchCategories = (request, response) => {
     })
 }
 
-const fetchReviews = () => {
-  
-    const reviewCommentCount = 
+const fetchReviews = (category, sort_by="created_at", order="desc") => {
+    const acceptedCategories = ["euro game", "social deduction", "dexterity", "children's games"];
```
Comment (Rosem10) [inline models/app.model.js]: This is not maintainable! What if a category is added to our database? What if your site really takes off and _thousands_ of categories are added? You'd need to add them in here manually each time. Luckily, though, if all of our categories are stored in our database, we could check them by utilising our existing GET categories endpoint.

File: models/app.model.js [hunk 6]
```diff
@@ -7,16 +7,34 @@ const fetchCategories = (request, response) => {
...
+
+    const acceptedSortBys = ["review_id","title", "designer", "owner", "review_img_url", "review_body", "category", "created_at", "votes"]
```
Comment (Rosem10) [inline models/app.model.js:17]: This, however, is okay to be hard coded like it is. All reviews will always follow this same format - as we have dictated in our seeding.

File: models/app.model.js [hunk 7]
```diff
@@ -7,16 +7,34 @@ const fetchCategories = (request, response) => {
...
-    GROUP BY reviews.review_id
-    ORDER BY created_at DESC;`;
+    LEFT JOIN comments ON comments.review_id=reviews.review_id`
     
-    return db.query(reviewCommentCount).then((result) => {
+    if(category !== undefined){
+        if(!acceptedCategories.includes(category.toLowerCase())){
+            return Promise.reject({status : 400, message : "Bad Request - invalid category filter"})
+        } else {
+            queryValues.push(category);
```
Comment (Rosem10) [inline models/app.model.js:30]: Lovely!

**Review by Rosem10 — CHANGES_REQUESTED** [review]
The code here is really strong, but the tests need a once-over please!

File: __tests__/app.test.js (hunk 2 above)
Comment (Wiggy93) [reply __tests__/app.test.js]: This is done by review id test on line 33 already so didn't want to repeat myself

File: __tests__/app.test.js (hunk 4 above)
Comment (Wiggy93) [reply __tests__/app.test.js]: Done in test on line 33

File: __tests__/app.test.js [hunk 8]
```diff
@@ -213,6 +213,54 @@ describe('PATCH commands', () => {
   
 });
 
+describe('Get /api/reviews Queries', () => {
+    test('200: should return the reviews with the category dexterity', () => {
```
Comment (Rosem10) [inline __tests__/app.test.js]: This needs to be rephrased as something like "should return the reviews of the category passed"

File: __tests__/app.test.js (hunk 3 above)
Comment (Rosem10) [reply __tests__/app.test.js]: Still needs the description updating, please.

File: __tests__/app.test.js (hunk 2 above)
Comment (Rosem10) [reply __tests__/app.test.js]: Fair enough!

File: __tests__/app.test.js [hunk 9]
```diff
@@ -213,6 +213,54 @@ describe('PATCH commands', () => {
...
+        return request(app).get("/api/reviews?sort_by=review_id&order=asc")
+        .expect(200)
+        .then(({body})=>{
+            const reviews = body.reviews
+            expect(reviews).toHaveLength(13);
+            expect(reviews).toBeSortedBy("review_id")
+        })
+    });
+
+    test('200: should return all reviews sorted by review_image_url descending', () => {
```
Comment (Rosem10) [inline __tests__/app.test.js:250]: Please rename this one too. Pop me a message if you need me to clarify - the descriptions should refer to the general behaviour, not it's behaviour in a specific instance.

File: models/app.model.js [hunk 10]
```diff
@@ -7,16 +7,39 @@ const fetchCategories = (request, response) => {
     })
 }
 
-const fetchReviews = () => {
-  
-    const reviewCommentCount = 
+const fetchReviews = (categories, category, sort_by="created_at", order="desc") => {
+    // const acceptedCategories = ["euro game", "social deduction", "dexterity", "children's games"];
```
Comment (Rosem10) [inline models/app.model.js:11]: Please remove!

File: controllers/app.controller.js [hunk 11]
```diff
@@ -1,3 +1,4 @@
+const categories = require("../db/data/test-data/categories")
```
Comment (Rosem10) [inline controllers/app.controller.js]: This needs changing please! What about when our data is no longer test data? You should make a request to the database for the categories. You have previously made a GET categories endpoint that you could use, which will be able to retrieve the categories from whichever db is currently being accessed.

**Review by Rosem10 — CHANGES_REQUESTED** [review]
This is better but there's still a few things to address please!

File: controllers/app.controller.js (hunk 11 above)
Comment (Wiggy93) [reply controllers/app.controller.js]: I didn't put this in, VS code must have automatically pulled it in and I missed it.

**Pull request merged**